<a href="https://colab.research.google.com/github/aarti-singh22/global-terrorism-analysis/blob/main/Global_Terrorism_by_Aarti.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> The Global Terrorism Database (GTD) is an open-source database including information on terrorist attacks around the world from 1970 through 2017. The GTD includes systematic data on domestic as well as international terrorist incidents that have occurred during this time period and now includes more than 180,000 attacks. The database is maintained by researchers at the National Consortium for the Study of Terrorism and Responses to Terrorism (START), headquartered at the University of Maryland.</b>

# <b> Explore and analyze the data to discover key findings pertaining to terrorist activities. </b>

**IMPORT LIBRARIES**

To get started we need to import some libraries.

In [ ]:
#Import the useful libraries.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
%matplotlib inline

**THE DATA**

Reading GLobal Terrorism Data.csv into Pandas dataframe.

In [ ]:
# Mount your drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Specifying file path 
filepath = '/content/drive/MyDrive/Global Terrorism Analysis – Aarti Hada/Copy of Global Terrorism Data.csv'
# To detect the character encoding of file.
import chardet
with open(filepath, 'rb') as rawdata:
    result = chardet.detect(rawdata.read(100000))
result

{'confidence': 0.73, 'encoding': 'ISO-8859-1', 'language': ''}

In [ ]:
# loading csv File
terrorism = pd.read_csv(filepath, encoding = 'ISO-8859-1')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (4,6,31,33,61,62,63,76,79,90,92,94,96,114,115,121) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [ ]:
# To show all columns of Pandas dataframe.
pd.set_option('display.max_columns', None)
# Print head of dataframe.
terrorism.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region                   region_txt provstate  \
0  Dominican Republic       2  Central America & Caribbean       NaN   
1              Mexico       1                North America   Federal   
2         Philippines       5               Southeast Asia    Tarlac   
3              Greece       8               Western Europe    Attica   
4               Japan       4                    East Asia   Fukouka   

            city   latitude   longitude  specificity  vicinity location  \
0  Santo Domingo  18.456792  -69.951164          1.0         0      NaN   
1    Mexico city  19.371887  -99.086624          1.0         0      NaN   
2        Unknown  15.478598  120.599741          4.0         0      NaN   
3         Athens  37.997490   23.762728          1.0         0      NaN   
4        Fukouka  33.580412  130.396361          1.0         0      NaN   

  summary  crit1  crit2  crit3  doubtterr  alternative alternative_txt  \
0     NaN      1      1      1        0.0          NaN             NaN   
1     NaN      1      1      1        0.0          NaN             NaN   
2     NaN      1      1      1        0.0          NaN             NaN   
3     NaN      1      1      1        0.0          NaN             NaN   
4     NaN      1      1      1       -9.0          NaN             NaN   

   multiple  success  suicide  attacktype1                 attacktype1_txt  \
0       0.0        1        0            1                   Assassination   
1       0.0        1        0            6     Hostage Taking (Kidnapping)   
2       0.0        1        0            1                   Assassination   
3       0.0        1        0            3               Bombing/Explosion   
4       0.0        1        0            7  Facility/Infrastructure Attack   

   attacktype2 attacktype2_txt  attacktype3 attacktype3_txt  targtype1  \
0          NaN             NaN          NaN             NaN         14   
1          NaN             NaN          NaN             NaN          7   
2          NaN             NaN          NaN             NaN         10   
3          NaN             NaN          NaN             NaN          7   
4          NaN             NaN          NaN             NaN          7   

                 targtype1_txt  targsubtype1  \
0  Private Citizens & Property          68.0   
1      Government (Diplomatic)          45.0   
2          Journalists & Media          54.0   
3      Government (Diplomatic)          46.0   
4      Government (Diplomatic)          46.0   

                                    targsubtype1_txt  \
0                                     Named Civilian   
1  Diplomatic Personnel (outside of embassy, cons...   
2                    Radio Journalist/Staff/Facility   
3                                  Embassy/Consulate   
4                                  Embassy/Consulate   

                         corp1                  target1  natlty1  \
0                          NaN             Julio Guzman     58.0   
1  Belgian Ambassador Daughter  Nadine Chaval, daughter     21.0   
2             Voice of America                 Employee    217.0   
3                          NaN             U.S. Embassy    217.0   
4                          NaN           U.S. Consulate    217.0   

          natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt  \
0  Dominican Republic        NaN           NaN           NaN              NaN   
1             Belgium        NaN           NaN           NaN              NaN   


In [ ]:
# Descriptive statistical details for the dataset
terrorism.describe()

eventid          iyear         imonth           iday  \
count  1.816910e+05  181691.000000  181691.000000  181691.000000   
mean   2.002705e+11    2002.638997       6.467277      15.505644   
std    1.325957e+09      13.259430       3.388303       8.814045   
min    1.970000e+11    1970.000000       0.000000       0.000000   
25%    1.991021e+11    1991.000000       4.000000       8.000000   
50%    2.009022e+11    2009.000000       6.000000      15.000000   
75%    2.014081e+11    2014.000000       9.000000      23.000000   
max    2.017123e+11    2017.000000      12.000000      31.000000   

            extended        country         region       latitude  \
count  181691.000000  181691.000000  181691.000000  177135.000000   
mean        0.045346     131.968501       7.160938      23.498343   
std         0.208063     112.414535       2.933408      18.569242   
min         0.000000       4.000000       1.000000     -53.154613   
25%         0.000000      78.000000       5.000000      11.510046   
50%         0.000000      98.000000       6.000000      31.467463   
75%         0.000000     160.000000      10.000000      34.685087   
max         1.000000    1004.000000      12.000000      74.633553   

          longitude    specificity       vicinity          crit1  \
count  1.771340e+05  181685.000000  181691.000000  181691.000000   
mean  -4.586957e+02       1.451452       0.068297       0.988530   
std    2.047790e+05       0.995430       0.284553       0.106483   
min   -8.618590e+07       1.000000      -9.000000       0.000000   
25%    4.545640e+00       1.000000       0.000000       1.000000   
50%    4.324651e+01       1.000000       0.000000       1.000000   
75%    6.871033e+01       1.000000       0.000000       1.000000   
max    1.793667e+02       5.000000       1.000000       1.000000   

               crit2          crit3      doubtterr   alternative  \
count  181691.000000  181691.000000  181690.000000  29011.000000   
mean        0.993093       0.875668      -0.523171      1.292923   
std         0.082823       0.329961       2.455819      0.703729   
min         0.000000       0.000000      -9.000000      1.000000   
25%         1.000000       1.000000       0.000000      1.000000   
50%         1.000000       1.000000       0.000000      1.000000   
75%         1.000000       1.000000       0.000000      1.000000   
max         1.000000       1.000000       1.000000      5.000000   

            multiple        success        suicide    attacktype1  \
count  181690.000000  181691.000000  181691.000000  181691.000000   
mean        0.137773       0.889598       0.036507       3.247547   
std         0.344663       0.313391       0.187549       1.915772   
min         0.000000       0.000000       0.000000       1.000000   
25%         0.000000       1.000000       0.000000       2.000000   
50%         0.000000       1.000000       0.000000       3.000000   
75%         0.000000       1.000000       0.000000       3.000000   
max         1.000000       1.000000       1.000000       9.000000   

       attacktype2  attacktype3      targtype1   targsubtype1        natlty1  \
count  6314.000000   428.000000  181691.000000  171318.000000  180132.000000   
mean      3.719512     5.245327       8.439719      46.971474     127.686441   
std       2.272023     2.246642       6.653838      30.953357      89.299120   
min       1.000000     1.000000       1.000000       1.000000       4.000000   
25%       2.000000     2.000000       3.000000      22.000000      83.000000   
50%       2.000000     7.000000       4.000000      35.000000     101.000000   
75%       7.000000     7.000000      14.000000      74.000000     173.000000   
max       9.000000     8.000000      22.000000     113.000000    1004.000000   

          targtype2  targsubtype2       natlty2    targtype3  targsubtype3  \
count  11144.000000  10685.000000  10828.000000  1176.000000   1097.000000   
mean      10.247218     55.311652    131.179442    1

In [ ]:
# Estimating the size of data
terrorism.shape

(181691, 135)

The dataset has 181691 rows and 135 columns.

# **EXPLORATORY DATA ANALYSIS**

We start with checking out the missing data.

**MISSING DATA**

Firstly, we will find out the null values and percentage of null values in each column.

In [ ]:
# To show all the rows of Pandas dataframe.
pd.set_option('display.max_rows', None)
# Checking missing values
total = terrorism.isnull().sum().sort_values(ascending = False)
# Percentage of Null values
percentage = ((terrorism.isnull().sum()/terrorism.isnull().count()) * 100 ).sort_values(ascending = False)
missing_values = pd.concat([total,percentage], axis = 1, keys = ['Total', 'Percentage'])
missing_values

Total  Percentage
gsubname3           181671   99.988992
weapsubtype4_txt    181621   99.961473
weapsubtype4        181621   99.961473
weaptype4           181618   99.959822
weaptype4_txt       181618   99.959822
claimmode3          181558   99.926799
claimmode3_txt      181558   99.926799
gsubname2           181531   99.911938
claim3              181373   99.824978
guncertain3         181371   99.823877
gname3              181367   99.821675
divert              181367   99.821675
attacktype3         181263   99.764435
attacktype3_txt     181263   99.764435
ransomnote          181177   99.717102
ransompaidus        181139   99.696187
ransomamtus         181128   99.690133
claimmode2          181075   99.660963
claimmode2_txt      181075   99.660963
ransompaid          180917   99.574002
corp3               180665   99.435305
targsubtype3        180594   99.396228
targsubtype3_txt    180594   99.396228
natlty3_txt         180544   99.368708
natlty3             180544   99.368708
target3             180516   99.353298
targtype3           180515   99.352747
targtype3_txt       180515   99.352747
ransomamt           180341   99.256980
weapsubtype3_txt    179998   99.068198
weapsubtype3        179998   99.068198
weaptype3_txt       179828   98.974633
weaptype3           179828   98.974633
claim2              179801   98.959772
guncertain2         179736   98.923997
gname2              179678   98.892075
resolution          179471   98.778145
kidhijcountry       178386   98.180978
nhours              177628   97.763786
compclaim           176852   97.336687
gsubname            175801   96.758232
attacktype2         175377   96.524869
attacktype2_txt     175377   96.524869
ndays               173567   95.528672
approxdate          172452   94.914993
corp2               171574   94.431755
nreleased           171291   94.275996
targsubtype2        171006   94.119136
targsubtype2_txt    171006   94.119136
natlty2             170863   94.040431
natlty2_txt         170863   94.040431
hostkidoutcome_txt  170700   93.950719
hostkidoutcome      170700   93.950719
target2             170671   93.934757
targtype2_txt       170547   93.866510
targtype2           170547   93.866510
weapsubtype2        170149   93.647456
weapsubtype2_txt    170149   93.647456
weaptype2           168564   92.775096
weaptype2_txt       168564   92.775096
nhostkidus          168174   92.560446
nhostkid            168119   92.530175
claimmode_txt       162608   89.497003
claimmode           162608   89.497003
related             156653   86.219461
addnotes            153402   84.430159
alternative         152680   84.032781
alternative_txt     152680   84.032781
propvalue           142702   78.541039
scite3              138175   76.049447
motive              131130   72.171984
location            126196   69.456385
propcomment         123732   68.100236
propextent          117626   64.739585
propextent_txt      117626   64.739585
scite2              104758   57.657231
ransom              104310   57.410659
nperps               71115   39.140629
nperpcap             69489   38.245703
nwoundte             69143   38.055270
weapdetail           67670   37.244553
nkillter             66958   36.852678
scite1               66191   36.430533
summary              66129   36.396409
claimed              66120   36.391456
nwoundus             64702   35.611010
nkillus              64446   35.470111
corp1                42550   23.418882
weapsubtype1         20768   11.430396
weapsubtype1_txt     20768   11.430396
nwound               16311    8.977330
targsubtype1         10373    5.709144
targsubtype1_txt     10373    5.709144
nkill                10313    5.676120
longitude             4557    2.508104
latitude              4556    2.507554
natlty1               1559    0.858050
natlty1_txt           1559    0.858050
target1                636    0.350045
city                   434    0.238867
provstate              421    0.231712
guncertain1            380    0.209146
isho

**Next, we will remove the columns with null values greater then 75% or the columns which does not contribute much to the analysis and keep only the columns which proffer to analysis.**

In [ ]:
# Keeping only necessary columns in the dataset
new_columns = ['success','attacktype1_txt', 'weaptype1_txt', 'region_txt', 'country_txt', 'gname', 'targtype1_txt', 'imonth', 'iday','iyear','provstate','target1','nkill','nwound','summary','motive' ]
terrorism = terrorism[new_columns]

In [ ]:
# Printing head of the modified dataset
terrorism.head()

success                 attacktype1_txt weaptype1_txt  \
0        1                   Assassination       Unknown   
1        1     Hostage Taking (Kidnapping)       Unknown   
2        1                   Assassination       Unknown   
3        1               Bombing/Explosion    Explosives   
4        1  Facility/Infrastructure Attack    Incendiary   

                    region_txt         country_txt  \
0  Central America & Caribbean  Dominican Republic   
1                North America              Mexico   
2               Southeast Asia         Philippines   
3               Western Europe              Greece   
4                    East Asia               Japan   

                                gname                targtype1_txt  imonth  \
0                              MANO-D  Private Citizens & Property       7   
1  23rd of September Communist League      Government (Diplomatic)       0   
2                             Unknown          Journalists & Media       1   
3                             Unknown      Government (Diplomatic)       1   
4                             Unknown      Government (Diplomatic)       1   

   iday  iyear provstate                  target1  nkill  nwound summary  \
0     2   1970       NaN             Julio Guzman    1.0     0.0     NaN   
1     0   1970   Federal  Nadine Chaval, daughter    0.0     0.0     NaN   
2     0   1970    Tarlac                 Employee    1.0     0.0     NaN   
3     0   1970    Attica             U.S. Embassy    NaN     NaN     NaN   
4     0   1970   Fukouka           U.S. Consulate    NaN     NaN     NaN   

  motive  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

In [ ]:
# Checking the shape to verify number of columns
terrorism.shape

(181691, 16)

The number of columns is reduced to 16.

**DATA DESCRIPTION**

To understand the data in the column and the meaning of the column name we will see the unique values of each column.


In [ ]:
# Returns all unique elements of the column success.
terrorism['success'].unique()

array([1, 0])

In [ ]:
# Returns all unique elements of the column attacktype1_txt.
terrorism['attacktype1_txt'].unique()

array(['Assassination', 'Hostage Taking (Kidnapping)',
       'Bombing/Explosion', 'Facility/Infrastructure Attack',
       'Armed Assault', 'Hijacking', 'Unknown', 'Unarmed Assault',
       'Hostage Taking (Barricade Incident)'], dtype=object)

In [ ]:
# Returns all unique elements of the column weaptype1_txt.
terrorism['weaptype1_txt'].unique()

array(['Unknown', 'Explosives', 'Incendiary', 'Firearms', 'Chemical',
       'Melee', 'Sabotage Equipment',
       'Vehicle (not to include vehicle-borne explosives, i.e., car or truck bombs)',
       'Fake Weapons', 'Radiological', 'Other', 'Biological'],
      dtype=object)

In [ ]:
# Returns all unique elements of the column region_txt.
terrorism['region_txt'].unique()

array(['Central America & Caribbean', 'North America', 'Southeast Asia',
       'Western Europe', 'East Asia', 'South America', 'Eastern Europe',
       'Sub-Saharan Africa', 'Middle East & North Africa',
       'Australasia & Oceania', 'South Asia', 'Central Asia'],
      dtype=object)

In [ ]:
# Returns all unique elements of the column country_txt.
terrorism['country_txt'].unique()

array(['Dominican Republic', 'Mexico', 'Philippines', 'Greece', 'Japan',
       'United States', 'Uruguay', 'Italy', 'East Germany (GDR)',
       'Ethiopia', 'Guatemala', 'Venezuela', 'West Germany (FRG)',
       'Switzerland', 'Jordan', 'Spain', 'Brazil', 'Egypt', 'Argentina',
       'Lebanon', 'Ireland', 'Turkey', 'Paraguay', 'Iran',
       'United Kingdom', 'Colombia', 'Bolivia', 'Nicaragua',
       'Netherlands', 'Belgium', 'Canada', 'Australia', 'Pakistan',
       'Zambia', 'Sweden', 'Costa Rica', 'South Yemen', 'Cambodia',
       'Israel', 'Poland', 'Taiwan', 'Panama', 'Kuwait',
       'West Bank and Gaza Strip', 'Austria', 'Czechoslovakia', 'India',
       'France', 'South Vietnam', 'Brunei', 'Zaire',
       "People's Republic of the Congo", 'Portugal', 'Algeria',
       'El Salvador', 'Thailand', 'Haiti', 'Sudan', 'Morocco', 'Cyprus',
       'Myanmar', 'Afghanistan', 'Peru', 'Chile', 'Honduras',
       'Yugoslavia', 'Ecuador', 'New Zealand', 'Malaysia', 'Singapore',
       'Bot

In [ ]:
# Returns all unique elements of the column gname.
terrorism['gname'].unique()

array(['MANO-D', '23rd of September Communist League', 'Unknown', ...,
       'Fatoni Warriors', 'Minorities of Metropolitan Attacks',
       'Baloch Republican Party'], dtype=object)

In [ ]:
# Returns all unique elements of the column targtype1_txt.
terrorism['targtype1_txt'].unique()

array(['Private Citizens & Property', 'Government (Diplomatic)',
       'Journalists & Media', 'Police', 'Utilities', 'Military',
       'Government (General)', 'Airports & Aircraft', 'Business',
       'Educational Institution', 'Violent Political Party',
       'Religious Figures/Institutions', 'Unknown', 'Transportation',
       'Tourists', 'NGO', 'Telecommunication', 'Food or Water Supply',
       'Terrorists/Non-State Militia', 'Other', 'Maritime',
       'Abortion Related'], dtype=object)

In [ ]:
# Returns all unique elements of the column imonth.
terrorism['imonth'].unique()

array([ 7,  0,  1,  2,  3,  4,  5,  6,  8,  9, 10, 11, 12])

In [ ]:
# Returns all unique elements of the column iday.
terrorism['iday'].unique()

array([ 2,  0,  1,  3,  6,  8,  9, 10, 11, 12, 13, 14, 15, 19, 20, 21, 22,
       25, 26, 27, 28, 30, 31,  4,  7, 16, 17, 18, 23, 24,  5, 29])

In [ ]:
# Returns all unique elements of the column iyear.
terrorism['iyear'].unique()

array([1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980,
       1981, 1986, 1982, 1983, 1984, 1985, 1987, 1988, 1989, 1990, 1991,
       1992, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003,
       2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014,
       2015, 2016, 2017])

In [ ]:
# Returns all unique elements of the column provstate.
terrorism['provstate'].unique()

array([nan, 'Federal', 'Tarlac', ..., 'Carlow', 'Western Lakes',
       'Vidzeme'], dtype=object)

In [ ]:
# Returns all unique elements of the column target1.
terrorism['target1'].unique()

array(['Julio Guzman', 'Nadine Chaval, daughter', 'Employee', ...,
       'Personal Security Officer of Council Member Ihuing Pame: Prafulla Phukan',
       'Jewelry Shop', 'Hmeymim Air Base'], dtype=object)

In [ ]:
# Returns all unique elements of the column nkill.
terrorism['nkill'].unique()

array([1.000e+00, 0.000e+00,       nan, 7.000e+00, 4.700e+01, 2.000e+00,
       3.600e+01, 5.000e+00, 3.000e+00, 4.000e+00, 2.500e+01, 1.500e+01,
       2.600e+01, 8.000e+00, 8.100e+01, 6.000e+00, 9.000e+00, 1.600e+01,
       3.000e+01, 3.100e+01, 1.200e+01, 2.100e+01, 1.400e+01, 8.800e+01,
       1.100e+01, 1.000e+01, 2.700e+01, 1.800e+01, 2.200e+01, 1.900e+01,
       9.200e+01, 1.300e+01, 7.300e+01, 1.000e+02, 4.200e+01, 4.300e+01,
       1.700e+01, 9.800e+01, 4.220e+02, 4.800e+01, 3.400e+01, 5.400e+01,
       5.000e+01, 3.500e+01, 2.000e+01, 4.100e+01, 3.700e+01, 2.800e+01,
       4.000e+01, 3.200e+01, 8.500e+01, 2.300e+01, 3.000e+02, 6.000e+01,
       2.400e+01, 5.800e+01, 7.000e+01, 8.700e+01, 4.500e+01, 3.800e+01,
       2.900e+01, 7.400e+01, 8.300e+01, 9.000e+01, 6.600e+01, 8.000e+01,
       6.700e+01, 5.100e+01, 3.900e+01, 1.140e+02, 1.240e+02, 7.600e+01,
       3.300e+01, 7.500e+01, 5.700e+01, 6.200e+01, 4.600e+01, 5.600e+01,
       6.300e+01, 1.200e+02, 1.020e+02, 7.800e+01, 

In [ ]:
# Returns all unique elements of the column nwound.
terrorism['nwound'].unique()

array([0.000e+00,       nan, 1.000e+00, 2.000e+00, 7.000e+00, 9.000e+00,
       1.700e+01, 5.000e+00, 3.000e+00, 1.000e+01, 1.200e+01, 2.000e+01,
       4.000e+00, 1.100e+01, 2.700e+01, 1.300e+01, 1.900e+01, 1.300e+02,
       5.600e+01, 2.400e+01, 6.000e+00, 7.200e+01, 2.380e+02, 5.500e+01,
       1.500e+01, 8.000e+00, 1.800e+01, 5.000e+01, 3.400e+01, 1.400e+01,
       7.000e+01, 1.020e+02, 4.100e+01, 4.800e+01, 3.600e+01, 2.500e+01,
       8.100e+01, 5.400e+01, 5.300e+01, 1.000e+02, 4.500e+01, 3.300e+01,
       4.000e+01, 6.200e+01, 7.400e+01, 2.200e+01, 3.100e+01, 1.600e+01,
       3.000e+01, 2.850e+02, 2.300e+01, 4.200e+01, 4.300e+01, 7.600e+01,
       4.400e+01, 5.200e+01, 4.600e+01, 4.900e+01, 1.600e+02, 3.700e+01,
       2.100e+01, 6.000e+01, 7.800e+01, 2.600e+01, 2.800e+01, 3.900e+01,
       3.800e+01, 3.500e+01, 3.200e+01, 8.000e+01, 4.700e+01, 1.880e+02,
       2.150e+02, 8.500e+01, 7.500e+01, 1.610e+02, 9.000e+01, 3.000e+02,
       9.500e+01, 1.350e+02, 6.100e+01, 1.380e+02, 

In [ ]:
# Returns all unique elements of the column summary.
terrorism['summary'].unique()

array([nan,
       '1/1/1970: Unknown African American assailants fired several bullets at police headquarters in Cairo, Illinois, United States.  There were no casualties, however, one bullet narrowly missed several police officers.  This attack took place during heightened racial tensions, including a Black boycott of White-owned businesses, in Cairo Illinois.',
       '1/2/1970: Unknown perpetrators detonated explosives at the Pacific Gas & Electric Company Edes substation in Oakland, California, United States.  Three transformers were damaged costing an estimated $20,000 to $25,000.  There were no casualties.',
       ...,
       '12/31/2017: Assailants set fire to houses in Kubentog, Datu Hoffer, Maguindanao, Philippines. There were no reported casualties in the attack. No group claimed responsibility for the incident; however, sources attributed the attack to the Bangsamoro Islamic Freedom Movement (BIFM).',
       '12/31/2017: Assailants threw a grenade at a Forest Department of

In [ ]:
# Returns all unique elements of the column motive.
terrorism['motive'].unique()

array([nan, 'To protest the Cairo Illinois Police Deparment',
       'To protest the War in Vietnam and the draft', ...,
       'The specific motive is unknown; however, sources stated that the attack was carried out in retaliation for the death of Noor Mohammad Tantray (Noor Trali), a Jaish-e-Mohammad (JeM) leader, by security personnel on December 26, 2017 in Pulwama.',
       'The specific motive is unknown; however, sources noted that the targeted building housed migrants.',
       'The specific motive is unknown; however, sources stated that the Taliban had warned one of the victims against collaborating with the government.'],
      dtype=object)

**COLUMN NAMES EXPLAINED**

1. success : The record of attack's success or failure..
2. attacktype1_txt :  The type of terrorist attack, whether it was Bombing, Explosion, Assassination etc.
3. weaptype1_txt : The type of weapons used by terrorist to attack.
4. region_txt : The region where attack happened.
5. country_txt : The country in that particular region where the attack took place.
6. gname : This refers to the terrorist group.
7. targtype1_txt : This refers to the group or media on which the attack was targeted.
8. imonth : month of attack.
9. iday : Date of attack.
10. iyear : year of attack.
11. provstate : Refers to the state.
12. target1 : Main target of attack.
13. nkill : Number of people killed.
14. nwound : Number of people wounded.
15. summary : Summary of the attack.
16. motive : Motive of the attack

In [ ]:
#Changing column names to make more sense.
terrorism.rename(columns = {'success':'Attack_Success','attacktype1_txt':'Attack_Type','weaptype1_txt':'Weapon_Used','region_txt':'Region','country_txt':'Country','gname':'Terrorist_group','targtype1_txt':'Targeted_group','imonth':'Month','iday':'Day','iyear':'Year','provstate':'State','target1':'Main_Target','nkill':'Killed','nwound':'Wounded','summary':'Attack_Summary','motive':'Attack_Motive'}, inplace = True)

In [ ]:
# Returns the column labels of the DataFrame
terrorism.columns

Index(['Attack_Success', 'Attack_Type', 'Weapon_Used', 'Region', 'Country',
       'Terrorist_group', 'Targeted_group', 'Month', 'Day', 'Year', 'State',
       'Main_Target', 'Killed', 'Wounded', 'Attack_Summary', 'Attack_Motive'],
      dtype='object')

**REMOVING DUPLICATE ROWS**

In [ ]:
#Shows number of duplicate rows
terrorism[terrorism.duplicated()].shape

(10039, 16)

In [ ]:
# Size of dataset
terrorism.shape

(181691, 16)

In [ ]:
#Remove Duplicates
terrorism.drop_duplicates(inplace= True)

In [ ]:
# Size of new dataset
terrorism.shape

(171652, 16)

**REMAINING MISSING DATAS**


In [ ]:
# Null values and its counts
terrorism.isnull().sum()

Attack_Success          0
Attack_Type             0
Weapon_Used             0
Region                  0
Country                 0
Terrorist_group         0
Targeted_group          0
Month                   0
Day                     0
Year                    0
State                 421
Main_Target           540
Killed               9263
Wounded             15091
Attack_Summary      58671
Attack_Motive      121790
dtype: int64

In [ ]:
# Concise summary of dataset
terrorism.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 171652 entries, 0 to 181690
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Attack_Success   171652 non-null  int64  
 1   Attack_Type      171652 non-null  object 
 2   Weapon_Used      171652 non-null  object 
 3   Region           171652 non-null  object 
 4   Country          171652 non-null  object 
 5   Terrorist_group  171652 non-null  object 
 6   Targeted_group   171652 non-null  object 
 7   Month            171652 non-null  int64  
 8   Day              171652 non-null  int64  
 9   Year             171652 non-null  int64  
 10  State            171231 non-null  object 
 11  Main_Target      171112 non-null  object 
 12  Killed           162389 non-null  float64
 13  Wounded          156561 non-null  float64
 14  Attack_Summary   112981 non-null  object 
 15  Attack_Motive    49862 non-null   object 
dtypes: float64(2), int64(4), object(10)
me

Fill the Null values.

In [ ]:
# Fill the missing values for Killed and Wounded with 0 and converting their datatypes from float to int.
terrorism['Killed'] = terrorism['Killed'].fillna(0).astype(int)
terrorism['Wounded'] = terrorism['Wounded'].fillna(0).astype(int)
# Fill the missing values for Attack Summary, Attack Motive and Main Target with Not Known
terrorism['Attack_Summary'].fillna('Not Known', inplace = True)
terrorism['Attack_Motive'].fillna('Not Known', inplace = True)
terrorism['Main_Target'].fillna('Not Known', inplace = True)

In [ ]:
# Checking Null values again
terrorism.isnull().sum()

Attack_Success       0
Attack_Type          0
Weapon_Used          0
Region               0
Country              0
Terrorist_group      0
Targeted_group       0
Month                0
Day                  0
Year                 0
State              421
Main_Target          0
Killed               0
Wounded              0
Attack_Summary       0
Attack_Motive        0
dtype: int64